File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/ui/playlist/ViewAdapter.java
package app.zxtune.ui.playlist;

import android.graphics.Rect;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.collection.LongSparseArray;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.selection.ItemDetailsLookup;
import androidx.recyclerview.selection.ItemKeyProvider;
import androidx.recyclerview.selection.Selection;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import app.zxtune.R;
import app.zxtune.databinding.PlaylistEntryBinding;

class ViewAdapter extends ListAdapter<Entry, ViewAdapter.EntryViewHolder> {

  interface Client {
    boolean onDrag(RecyclerView.ViewHolder holder);
  }

  private final Client client;
  private final LongSparseArray<Integer> positionsCache;
  @Nullable
  private List<Entry> list;
  @Nullable
  private Selection<Long> selection;
  private boolean isPlaying = false;
  @Nullable
  private Long nowPlaying = null;
  @Nullable
  private Integer nowPlayingPos = null;

  ViewAdapter(Client client) {
    super(new DiffUtil.ItemCallback<Entry>() {
      @Override
      public boolean areItemsTheSame(Entry oldItem, Entry newItem) {
        return oldItem.getId() == newItem.getId();
      }

      @Override
      public boolean areContentsTheSame(Entry oldItem, Entry newItem) {
        return TextUtils.equals(oldItem.getTitle(), newItem.getTitle())
            && TextUtils.equals(oldItem.getAuthor(), newItem.getAuthor())
            && oldItem.getDuration().equals(newItem.getDuration());
      }
    });
    this.client = client;
    this.positionsCache = new LongSparseArray<>();
    setHasStableIds(true);
  }

  final void setSelection(@Nullable Selection<Long> selection) {
    this.selection = selection;
  }

  final void setIsPlaying(boolean isPlaying) {
    if (this.isPlaying != isPlaying) {
      updateNowPlaying();
      this.isPlaying = isPlaying;
    }
  }

  final void setNowPlaying(@Nullable Long id) {
    if (nowPlaying != null && !nowPlaying.equals(id)) {
      updateNowPlaying();
    }
    nowPlaying = id;
    nowPlayingPos = getPosition(nowPlaying);
    updateNowPlaying();
  }

  private void updateNowPlaying() {
    if (nowPlayingPos != null) {
      notifyItemChanged(nowPlayingPos);
    }
  }

  @Nullable
  private Integer getPosition(@Nullable Long id) {
    if (id == null) {
      return null;
    }
    final Integer cached = positionsCache.get(id);
    if (cached != null) {
      return cached;
    }
    //TODO: lookup?
    final long numId = id;
    for (int pos = positionsCache.size(), lim = getItemCount(); pos < lim; ++pos) {
      final long itemId = getItem(pos).getId();
      positionsCache.append(itemId, pos);
      if (numId == itemId) {
        return pos;
      }
    }
    return null;
  }

  final void onItemMove(int fromPosition, int toPosition) {
    if (fromPosition == toPosition) {
      return;
    }
    if (fromPosition < toPosition) {
      moveForward(fromPosition, toPosition);
    } else {
      moveBackward(fromPosition, toPosition);
    }
    notifyItemMoved(fromPosition, toPosition);
  }

  private void moveForward(int fromPosition, int toPosition) {
    final Entry moved = list.get(fromPosition);
    for (int pos = fromPosition; pos < toPosition; ++pos) {
      placeItem(pos, list.get(pos + 1));
    }
    placeItem(toPosition, moved);
  }

  private void moveBackward(int fromPosition, int toPosition) {
    final Entry moved = list.get(fromPosition);
    for (int pos = fromPosition; pos > toPosition; --pos) {
      placeItem(pos, list.get(pos - 1));
    }
    placeItem(toPosition, moved);
  }

  private void placeItem(int pos, Entry entry) {
    list.set(pos, entry);
    positionsCache.put(entry.getId(), pos);
  }

  @Override
  public void submitList(@Nullable List<Entry> list) {
    this.list = list;
    this.positionsCache.clear();
    super.submitList(list);
  }

  @Override
  public long getItemId(int position) {
    return getItem(position).getId();
  }

  @Override
  public EntryViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
    final PlaylistEntryBinding binding = DataBindingUtil.inflate(inflater, R.layout.playlist_entry, parent, false);
    return new EntryViewHolder(binding);
  }

  @Override
  public void onBindViewHolder(final EntryViewHolder holder, int position) {
    final Entry entry = getItem(position);
    holder.bind(entry, isPlaying && isNowPlaying(entry.getId()), isSelected(entry.getId()));
  }

  private boolean isNowPlaying(long id) {
    return nowPlaying != null && nowPlaying == id;
  }

  private boolean isSelected(long id) {
    return selection != null && selection.contains(id);
  }

  @Override
  public void onViewAttachedToWindow(final EntryViewHolder holder) {
    holder.binding.playlistEntryState.setOnTouchListener((v, event) -> {
      if (event.getAction() == MotionEvent.ACTION_DOWN) {
        return client.onDrag(holder);
      }
      return false;
    });
  }

  @Override
  public void onViewDetachedFromWindow(EntryViewHolder holder) {
    holder.binding.playlistEntryState.setOnTouchListener(null);
  }

  static class EntryViewHolder extends RecyclerView.ViewHolder {
    private final PlaylistEntryBinding binding;

    EntryViewHolder(PlaylistEntryBinding binding) {
      super(binding.getRoot());
      this.binding = binding;
    }

    final void bind(Entry entry, boolean isPlaying, boolean isSelected) {
      binding.setEntry(entry);
      binding.setIsPlaying(isPlaying);
      binding.executePendingBindings();
      itemView.setSelected(isSelected);
    }
  }

  static class HolderItemDetails extends ItemDetailsLookup.ItemDetails<Long> {

    private final RecyclerView.ViewHolder holder;

    HolderItemDetails(RecyclerView.ViewHolder holder) {
      this.holder = holder;
    }

    @Override
    public int getPosition() {
      return holder.getAdapterPosition();
    }

    @Nullable
    @Override
    public Long getSelectionKey() {
      final long id = holder.getItemId();
      return id != RecyclerView.NO_ID ? id : null;
    }
  }

  static class KeyProvider extends ItemKeyProvider<Long> {

    private final ViewAdapter adapter;

    KeyProvider(ViewAdapter adapter) {
      super(SCOPE_MAPPED);
      this.adapter = adapter;
    }

    @Nullable
    @Override
    public Long getKey(int position) {
      return adapter.getItem(position).getId();
    }

    @Override
    public int getPosition(Long key) {
      final Integer pos = adapter.getPosition(key);
      return pos != null ? pos : RecyclerView.NO_POSITION;
    }
  }

  static class DetailsLookup extends ItemDetailsLookup<Long> {

    private final RecyclerView listing;

    DetailsLookup(RecyclerView view) {
      this.listing = view;
    }

    @Nullable
    @Override
    public ItemDetails<Long> getItemDetails(MotionEvent e) {
      float x = e.getX();
      float y = e.getY();
      final View item = listing.findChildViewUnder(x, y);
      if (item != null) {
        final EntryViewHolder holder = (EntryViewHolder) listing.getChildViewHolder(item);
        final Rect rect = new Rect();
        holder.binding.playlistEntryState.getHitRect(rect);
        x -= item.getX();
        y -= item.getY();
        // return detail only if not d'n'd event
        return rect.contains((int) x, (int) y)
            ? null
            : new HolderItemDetails(holder);
      }
      return null;
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/ui/playlist/ViewAdapter.kt
package app.zxtune.ui.playlist

import android.annotation.SuppressLint
import android.graphics.Rect
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.ViewGroup
import androidx.collection.LongSparseArray
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.selection.ItemDetailsLookup
import androidx.recyclerview.selection.ItemDetailsLookup.ItemDetails
import androidx.recyclerview.selection.ItemKeyProvider
import androidx.recyclerview.selection.Selection
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import app.zxtune.R
import app.zxtune.databinding.PlaylistEntryBinding
import app.zxtune.playlist.IdType
import app.zxtune.playlist.Track

internal class ViewAdapter(private val client: Client) :
    ListAdapter<Track, ViewAdapter.TrackViewHolder>(DiffCallback()) {
    internal fun interface Client {
        fun move(track: Track.Id, delta: Int)
    }

    init {
        setHasStableIds(true)
    }

    private val positionsCache: LongSparseArray<Int> = LongSparseArray()
    private var mutableList: MutableList<Track>? = null
    private lateinit var selection: Selection<IdType>
    private lateinit var touchHelper: CustomTouchHelper
    private var isPlaying = false
    private var nowPlaying: Track.Id? = null
    private var nowPlayingPos: Int? = null

    fun setSelection(selection: Selection<IdType>) {
        this.selection = selection
    }

    fun setIsPlaying(isPlaying: Boolean) {
        if (this.isPlaying != isPlaying) {
            updateNowPlaying()
            this.isPlaying = isPlaying
        }
    }

    fun setNowPlaying(id: Track.Id?) {
        if (nowPlaying != null && nowPlaying != id) {
            updateNowPlaying()
        }
        nowPlaying = id
        nowPlayingPos = id?.let { getPosition(it.value) }
        updateNowPlaying()
    }

    private fun updateNowPlaying() = nowPlayingPos?.let {
        notifyItemChanged(it)
    } ?: Unit

    private fun getPosition(id: IdType): Int? {
        positionsCache[id]?.let {
            return it
        }
        //TODO: lookup?
        for (pos in positionsCache.size() until itemCount) {
            val itemId = getItem(pos).id
            positionsCache.append(itemId.value, pos)
            if (id == itemId.value) {
                return pos
            }
        }
        return null
    }

    fun onItemMove(fromPosition: Int, toPosition: Int) {
        if (fromPosition == toPosition) {
            return
        }
        val list = mutableList ?: return
        if (maxOf(fromPosition, toPosition) >= list.size) {
            return
        }
        val placeItem = { pos: Int, track: Track ->
            list[pos] = track
            positionsCache.put(track.id.value, pos)
        }
        val moved = list[fromPosition]
        if (fromPosition < toPosition) {
            for (pos in fromPosition until toPosition) {
                placeItem(pos, list[pos + 1])
            }
        } else {
            for (pos in fromPosition downTo toPosition + 1) {
                placeItem(pos, list[pos - 1])
            }
        }
        placeItem(toPosition, moved)
        notifyItemMoved(fromPosition, toPosition)
    }

    // TODO: think about another solution about D&D detection
    override fun submitList(list: List<Track>?, callback: Runnable?) {
        if (touchHelper.isDragging) {
            callback?.run()
        } else {
            positionsCache.clear()
            mutableList = list as? MutableList<Track>
            super.submitList(list, callback)
        }
    }

    override fun getItemId(position: Int) = getItem(position).id.value

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TrackViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding: PlaylistEntryBinding =
            DataBindingUtil.inflate(inflater, R.layout.playlist_entry, parent, false)
        return TrackViewHolder(binding)
    }

    override fun onBindViewHolder(holder: TrackViewHolder, position: Int) {
        getItem(position).let { entry ->
            holder.bind(entry, isPlaying && isNowPlaying(entry.id), isSelected(entry.id))
        }
    }

    private fun isNowPlaying(id: Track.Id) = id == nowPlaying

    private fun isSelected(id: Track.Id) = selection.contains(id.value)

    private fun hasSelection() = !selection.isEmpty

    @SuppressLint("ClickableViewAccessibility")
    override fun onViewAttachedToWindow(holder: TrackViewHolder) {
        holder.binding.playlistEntryState.setOnTouchListener { _, event ->
            if (event.action == MotionEvent.ACTION_DOWN && !hasSelection() && mutableList != null) {
                touchHelper.startDrag(holder)
                true
            } else {
                false
            }
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onViewDetachedFromWindow(holder: TrackViewHolder) =
        holder.binding.playlistEntryState.setOnTouchListener(null)

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        touchHelper = CustomTouchHelper(TouchHelperCallback()).apply {
            attachToRecyclerView(recyclerView)
        }
    }

    private class CustomTouchHelper(private val callback: TouchHelperCallback) :
        ItemTouchHelper(callback) {

        val isDragging
            get() = callback.isDragging
    }

    private inner class TouchHelperCallback :
        ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP or ItemTouchHelper.DOWN, 0) {

        private var draggedItem: Track.Id? = null
        private var dragDelta = 0

        val isDragging
            get() = draggedItem != null

        override fun isLongPressDragEnabled() = false

        override fun isItemViewSwipeEnabled() = false

        override fun onSelectedChanged(
            viewHolder: RecyclerView.ViewHolder?,
            actionState: Int
        ) {
            if (actionState != ItemTouchHelper.ACTION_STATE_IDLE) {
                viewHolder?.apply {
                    itemView.isActivated = true
                }
            }
            super.onSelectedChanged(viewHolder, actionState)
        }

        override fun clearView(view: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
            super.clearView(view, viewHolder)
            viewHolder.itemView.isActivated = false
            draggedItem.takeIf { dragDelta != 0 }?.let {
                client.move(it, dragDelta)
            }
            draggedItem = null
            dragDelta = 0
        }

        override fun onMove(
            recyclerView: RecyclerView,
            source: RecyclerView.ViewHolder,
            target: RecyclerView.ViewHolder
        ): Boolean {
            val srcPos = source.bindingAdapterPosition
            val tgtPos = target.bindingAdapterPosition
            if (draggedItem == null) {
                draggedItem = Track.Id(source.itemId)
            }
            dragDelta += tgtPos - srcPos
            onItemMove(srcPos, tgtPos)
            return true
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) = Unit
    }

    internal class TrackViewHolder(val binding: PlaylistEntryBinding) :
        RecyclerView.ViewHolder(binding.root) {
        fun bind(track: Track, isPlaying: Boolean, isSelected: Boolean) {
            binding.meta = track.meta
            binding.isPlaying = isPlaying
            binding.executePendingBindings()
            itemView.isSelected = isSelected
        }
    }

    internal class HolderItemDetails(private val holder: RecyclerView.ViewHolder) :
        ItemDetails<IdType>() {
        override fun getPosition() = holder.bindingAdapterPosition

        override fun getSelectionKey() = holder.itemId
    }

    internal class KeyProvider(private val adapter: ViewAdapter) :
        ItemKeyProvider<IdType>(SCOPE_MAPPED) {
        override fun getKey(position: Int) = adapter.getItem(position).id.value

        override fun getPosition(key: IdType) = adapter.getPosition(key) ?: RecyclerView.NO_POSITION
    }

    internal class DetailsLookup(
        private val listing: RecyclerView,
        private val adapter: ViewAdapter
    ) : ItemDetailsLookup<IdType>
        () {
        override fun getItemDetails(e: MotionEvent): ItemDetails<IdType>? {
            if (adapter.touchHelper.isDragging) {
                return null
            }
            var x = e.x
            var y = e.y
            val item = listing.findChildViewUnder(x, y)
            if (item != null) {
                val holder = listing.getChildViewHolder(item) as TrackViewHolder
                val rect = Rect()
                holder.binding.playlistEntryState.getHitRect(rect)
                x -= item.x
                y -= item.y
                // return detail only if not d'n'd event
                return if (rect.contains(x.toInt(), y.toInt())) null else HolderItemDetails(holder)
            }
            return null
        }
    }

    private class DiffCallback : DiffUtil.ItemCallback<Track>() {
        override fun areItemsTheSame(oldItem: Track, newItem: Track) = oldItem.id == newItem.id

        override fun areContentsTheSame(oldItem: Track, newItem: Track) = oldItem == newItem
    }
}